### ISSUE TITLE
IDbContextFactory<TContext> is not covariant

### ISSUE THREAD

**Issue by Driedas** [dotnet/efcore#30096]
According to the [official documentation of IDbContextFactory<TContext>](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.idbcontextfactory-1?view=efcore-7.0), the TContext parameter should be covariant:

>TContext
The [DbContext](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext?view=efcore-7.0) type to create.
This type parameter is covariant. That is, you can use either the type you specified or any type that is more derived. For more information about covariance and contravariance, see [Covariance and Contravariance in Generics](https://learn.microsoft.com/en-us/dotnet/standard/generics/covariance-and-contravariance).

However, that is not the case in the [source code](https://github.com/dotnet/efcore/blob/main/src/EFCore/IDbContextFactory.cs#LL15C3-L15C3)
```
public interface IDbContextFactory<[DynamicallyAccessedMembers(DbContext.DynamicallyAccessedMemberTypes)] TContext>
    where TContext : DbContext
{
```

Is this just a case of a copy paste issue from the documentation of Entity Framework, where the interface actually _was_ covariant?
Is there a reason why this is not supported in EF core? To me it looks like a textbook example of where covariance should work and enable to dynamically provide derived DbContext implementations with different configurations. For example we are using a derived DbContext in automated tests with certain entities configured slightly differently (e. g. sql server variant columns)

TLDR: The below code should work according to the documentation, and did work in old EF, but does not in EF core
```
IDbContextFactory<DbContextBase> contextFactory = _serviceProvider.GetRequiredService<IDbContextFactory<DbContextDerived>>();
```

**Comment by roji** [comment]
The question was raised #26630; the interface used to be covariant, but adding the async method removed that. We should fix the docs.

**Comment by Driedas** [comment]
right, thanks for the quick response, I did try looking for existing issues, but probably only open issues by mistake 👍

In the mean time I've implemented a workaround, coincidentally the same approach used in the linked issue, so we will stick with that.

Thanks...

**Comment by KRamey-UPS** [comment]
Workaround:
```
public class DbFactoryCovarianceAdapter<TBaseContext, TSubContext>(IDbContextFactory<TSubContext> concreteFactory) 
  : IDbContextFactory<TBaseContext>
    where TBaseContext : DbContext
    where TSubContext : TBaseContext
{
  public TBaseContext CreateDbContext() 
    => concreteFactory.CreateDbContext();
  public async Task<TBaseContext> CreateDbContextAsync(CancellationToken cancellationToken = default)
      => await concreteFactory.CreateDbContextAsync(cancellationToken);
}
```
